### PR TITLE
Migrate multi_token_prediction to NNX

### PR DESCRIPTION
**Pull request by ecnal-cienet**
# Description

Migrate Multi_Token_Prediction (MTP) to use NNX module.

# Tests

We use [xpk](https://github.com/AI-Hypercomputer/xpk) to create tpu cluster and assign workload.

# Environment

Machine type : `v5litepod-8` TPU VM
Software machine : `tpu-ubuntu2204-base`
Python Version : `3.12.11`

# Test Command
Run train command to train qwen3-0.6b for 25 steps:
```
python3 -m MaxText.train MaxText/configs/base.yml run_name=mtp_linen_validation_run base_output_directory=gs://lance-maxtext/maxtext_output model_name=qwen3-0.6b dataset_type=synthetic steps=25 per_device_batch_size=1 metrics_file=metrics.txt mtp_num_layers=8 mtp_loss_scaling_factor=0.1 mtp_eval_target_module=1 async_checkpointing=false max_target_length=256 ici_pipeline_parallelism=4
```

Logs:
[Linen, before migration](https://paste.googleplex.com/6212477348085760)
[NNX, after migration](https://paste.googleplex.com/5365437051305984)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Comment by parambole** [comment]
> @parambole do you know the best way to test `multi_token_prediction`? Since you are the original file author

Hey @bvandermoon - Only we tested it using two mechanisms - Setting MTP enabled
1. Doing a training run on a default model and observing the MTP and main model loss
2. Loading the deepseek checkpoints and running a couple of fine tuning steps. To confirm the base MTP and main model loss and track it for a few steps.


cc: @RissyRan @Rohan-Bierneni

File: src/MaxText/layers/multi_token_prediction.py [hunk 1]
```diff
@@ -174,22 +185,40 @@ def __call__(
...
+    self.rngs = rngs if rngs is not None else nnx.Rngs(0)
+    for k in range(1, config.mtp_num_layers + 1):
+      layer = MultiTokenPredictionLayer(
+        config=config,
+        mesh=mesh,
+        layer_number=k,
+        transformer_layer_module=transformer_layer_module,
+        rngs=rngs,
+      )
+      setattr(self, f"mtp_layer_{k}", layer)
```
Comment (ecnal-cienet) [inline src/MaxText/layers/multi_token_prediction.py:202]: I find this part a bit confusing for other reviewers.
In my previous version, the code looked like this:
```
self.mtp_layers = nnx.List([])
    for k in range(1, config.mtp_num_layers + 1):
      layer = MultiTokenPredictionLayer(
        config=config,
        mesh=mesh,
        layer_number=k,
        transformer_layer_module=transformer_layer_module,
        rngs=rngs,
      )
      setattr(self, f"mtp_layer_{k}", layer)
      self.mtp_layers.append(layer)
```
This worked fine locally and all test suites passed. However, the `tpu_unit_tests / run` check reported the following error:
```
flax.nnx has no attribute List
```
So I’m not entirely sure whether we should keep the original approach (since it worked in my environment) or stick with the current version for broader compatibility.

File: src/MaxText/layers/models.py [hunk 2]
```diff
@@ -92,17 +92,23 @@ def setup(self):
...
+      mtp_block = nnx_wrappers.to_linen_class(
+          base_nnx_class=MultiTokenPredictionBlock,
+          base_metadata_fn=initializers.variable_to_logically_partitioned,
+      )
+      self.mtp_block = mtp_block(
+          config=self.config,
+          mesh=self.mesh,
+          transformer_layer_module=mtp_layer,
+          decoder=self.decoder,
       )
```
Comment (bvandermoon) [inline src/MaxText/layers/models.py]: Instead of using `to_linen_class`, could you create a method in `multi_token_prediction.py` using `nnx.bridge.to_linen` and call it here?

* [Example method creation](https://github.com/AI-Hypercomputer/maxtext/blob/31af45d0cc91eb54fdf42b61c56262428b9056b0/MaxText/layers/embeddings.py#L74-L85)
* [Example usage](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/MaxText/layers/models.py#L77-L85)

Comment (ecnal-cienet) [reply src/MaxText/layers/models.py]: Hi @bvandermoon , thanks for the comment. I have refactored the code. Please review it.

Thanks.

**Review by RissyRan — COMMENTED** [review]
Thanks for the change!

I see the loss is much larger: https://screenshot.googleplex.com/9Qcth6EFDuKTBnm

Could you check if anything wrong during migration?

**Comment by ecnal-cienet** [comment]
> Thanks for the change!
> 
> I see the loss is much larger: https://screenshot.googleplex.com/9Qcth6EFDuKTBnm
> 
> Could you check if anything wrong during migration?

Hi @RissyRan , thanks for the comment. I have fixed the missing `lazy_init` which caused the slower loss convergence issue in NNX code.

File: src/MaxText/layers/models.py [hunk 3]
```diff
@@ -94,16 +94,20 @@ def setup(self):
...
+          rngs=self.make_rng("mtp_block"),
       )
 
   def logits_from_hidden_states(self, hidden_states, deterministic, model_mode):
     """
     Compute logits from hidden states (wrapping decoder.apply_output_head).
     This function is only used for vocabulary tiling.
     """
-    logits = self.decoder.apply_output_head(
+    logits = self.decoder.apply_output_head(  # pylint: disable=protected-access
```
Comment (NicoGrande) [inline src/MaxText/layers/models.py]: The pylint line shouldn't be necessary anymore.

Comment (ecnal-cienet) [reply src/MaxText/layers/models.py]: Done. Thanks.
